samcli/lib/telemetry/metric.py

```diff
@@ -29,16 +30,120 @@
 No side effect will result in this as it is write-only for code outside of telemetry.
 Decorators should be used to minimize logic involving telemetry.
 """
-_METRICS = dict()
+_METRICS: Optional[defaultdict] = None
+
+
+class MetricName(Enum):
+    # Metrics that can be reported only once in a session
+    installed = ("installed", True)
+    commandRun = ("commandRun", True)
+
+    # Metrics that can be reported any number of times during a session
+    # e.g. docker_image_download_status
+    templateWarning = ("templateWarning", False)
+    runtimeMetric = ("runtimeMetric", False)
+
+    def __init__(self, name: str, can_report_once_only: bool):
+        self._name = name
+        self.can_report_once_only = can_report_once_only
+
+    def __str__(self) -> str:
+        return self._name
+
+
+def _get_global_metric_sink() -> defaultdict:
+    global _METRICS  # pylint: disable=global-statement
+    if _METRICS is None:
+        _METRICS = defaultdict(list)
+    return _METRICS
+
+
+def flush():
+    """
+    Emits all metrics to telemetry and flush the metric sink
+    """
+    telemetry = Telemetry()
+    metric_sink = _get_global_metric_sink()
+    for metric_list in metric_sink.values():
+        while metric_list:
+            metric = metric_list.pop(0)
+            telemetry.emit(metric)
+
+
+def track_metric(
+    metric_name: MetricName,
+    metric_specific_attrs: Optional[Dict[str, Any]] = None,
+    overwrite_existing_metric_specific_attrs: bool = False,
+    top_level_attrs: Optional[Dict[str, Any]] = None,
+    overwrite_existing_top_level_attrs: bool = False,
+):
+    """
+    Generic metric tracker
+    Metrics tracked by Generic metric tracker will first be collected into the global metric sink,
+    All metrics collected in the global metric sink will be emitted at program exit
+    (by calling flush(), see samcli/cli/main.py)
+
+    Examples
+    --------
+    tracking metric with metric-specific attributes
+    >>> track_metric(MetricName.commandRun, {"foo": "bar"})
+    >>> track_metric(MetricName.commandRun, metric_specific_attrs={"foo": "bar"})
+
+    overwriting exisitng metric-specific attributes
```

Comment:
Would we ever over-write, what are the conditions under which this would be used, I notice we also mention a note of caution while using these APIs, how destructive are they?

Reply:
In the `track_command` decorator, certain attributes are added (e.g. `region`, `duration`) after the main command behaviors complete. Current `track_metric` method allows the possibility that these attributes to be added prematurely (e.g. when `track_metric` is called somewhere during the command session). 

Thinking a bit deeper, I think we should have a list of "protected attributes" that can be added by `track_metric`